### PR TITLE
Fix beam search implementation

### PR DESCRIPTION
**Pull request by huzecong**
### Problem
- When "remove_duplicates_in_prediction" is set to true, candidate beams are discarded if it contains duplicate var names. When all candidates are discarded, no results are generated for an example. This is true of ~1.6% of examples.

### Solution
- We can prevent this by disallowing generation of <EOT> if it would result in duplicate names.
- A "workaround" that produces the same results without modifying logits is: instead of exploring only the top-k candidates, take the top-(k + len(remaining_beams)) ones. This way, even if the top-ranking candidates are all invalid <EOT> ones (one from each beam), after discarding them, we still get k candidates.

### Result
- This fix resulted in a ~0.1% increase in accuracy on the test set without retraining the model. Only ~0.01% examples yielded no results.

### CODE REVIEW

**Comment by pcyin** [comment]
Thanks!
